test/expect-tests/ocamlc_loc/ocamlc_loc_tests.ml

```diff
@@ -0,0 +1,168 @@
+open Stdune
+
+let cmd fmt =
+  Printf.ksprintf
+    (fun s ->
+      let (_ : int) = Sys.command s in
+      ())
+    fmt
+
+module Test = struct
+  type t = { dir : Path.t }
+
+  let restore_cwd =
+    let cwd = Sys.getcwd () in
+    fun () -> Sys.chdir cwd
+
+  let file t ~fname ~contents =
+    let path = Path.relative t.dir fname in
+    Io.write_file path contents;
+    path
+
+  let string_of_loc { Ocamlc_loc.path; line; chars } =
+    sprintf "file %s, line %s%s" path
+      (match line with
+      | `Single i -> string_of_int i
+      | `Range (x, y) -> sprintf "%d-%d" x y)
+      (match chars with
+      | None -> ""
+      | Some (x, y) -> sprintf ", chars %d-%d" x y)
+
+  let print_message message =
+    match (message : Ocamlc_loc.message) with
+    | Raw s -> printfn "raw message =\n%s" s
+    | Structured { preview; message; severity } ->
+      printfn "severity = %s"
+        (match severity with
+        | Error -> "error"
+        | Warning { code; name } -> sprintf "warning %d [%s]" code name);
+      Option.iter preview ~f:(printfn "preview =\n%s");
+      printfn "message = %s" message
+
+  let create f =
+    let dir = Temp.create Dir ~prefix:"dune." ~suffix:".test" in
+    let t = { dir } in
+    Sys.chdir (Path.to_string dir);
+    let out_file = Exn.protect ~f:(fun () -> f t) ~finally:restore_cwd in
+    let output = Io.read_file out_file in
+    (* Format.eprintf "print raw output:@.%s@.%!" output; *)
+    let locs =
+      Ocamlc_loc.parse
+        (Format.asprintf "%a@." Pp.to_fmt (Ansi_color.parse output))
+    in
+    List.iteri locs ~f:(fun i { Ocamlc_loc.loc; message; related } ->
+        printfn ">> error %d" i;
+        printfn "loc = %s" (string_of_loc loc);
+        print_message message;
+        if related <> [] then (
+          printfn "related errors:";
+          List.iter related ~f:(fun (loc, message) ->
+              print_message message;
+              printfn "loc = %s" (string_of_loc loc))
+        ))
+end
+
+let%expect_test "" =
+  Test.create (fun t ->
+      let open Test in
+      let (_ : Path.t) = file t ~fname:"test.ml" ~contents:"let () = 123" in
+      cmd "ocamlc -c test.ml 2> out";
+      Path.relative t.dir "out");
+  [%expect
+    {|
+    >> error 0
+    loc = file test.ml, line 1, chars 9-12
```

Comment:
IMO, printing a `Dyn.t` would be clearer here. It would avoid having to understand the printer to understand what exact value the parser captured.

Reply:
Changed it to use `Dyn.t`